### ISSUE TITLE
cEntity:Destroy: Verify failed

### ISSUE THREAD

**Issue by Seadragon91** [cuberite/cuberite#3994]
Server OS:        Windows

### Actual behavior
Server crashes:
```
 	Cuberite.exe!abort() Line 77
 	Cuberite.exe!NonCtrlHandler(int a_Signal) Line 107
 	Cuberite.exe!raise(int signum) Line 516
 	Cuberite.exe!abort() Line 71
 	Cuberite.exe!terminate() Line 58
 	[External Code]		Annotated Frame
 	Cuberite.exe!_execute_onexit_table::__l2::<lambda>() Line 206
 	Cuberite.exe!__crt_seh_guarded_call<int>::operator()<void <lambda>(void),int <lambda>(void) & __ptr64,void <lambda>(void) >(__acrt_lock_and_call::__l2::void <lambda>(void) && setup, _execute_onexit_table::__l2::int <lambda>(void) & action, __acrt_lock_and_call::__l2::void <lambda>(void) && cleanup) Line 204
 	[Inline Frame] Cuberite.exe!__acrt_lock_and_call(const __acrt_lock_id) Line 908
 	Cuberite.exe!_execute_onexit_table(_onexit_table_t * table) Line 231
 	Cuberite.exe!common_exit::__l2::<lambda>() Line 215
 	Cuberite.exe!__crt_seh_guarded_call<void>::operator()<void <lambda>(void),void <lambda>(void) & __ptr64,void <lambda>(void) >(__acrt_lock_and_call::__l2::void <lambda>(void) && setup, common_exit::__l2::void <lambda>(void) & action, __acrt_lock_and_call::__l2::void <lambda>(void) && cleanup) Line 224
 	[Inline Frame] Cuberite.exe!__acrt_lock_and_call(const __acrt_lock_id) Line 908
 	Cuberite.exe!common_exit(const int return_code, const _crt_exit_cleanup_mode cleanup_mode, const _crt_exit_return_mode return_mode) Line 263
>	[Inline Frame] Cuberite.exe!cEntity::Destroy::__l2::<lambda>(cWorld &) Line 251
 	[Inline Frame] Cuberite.exe!std::_Invoker_functor::_Call(cEntity::Destroy::__l2::void <lambda>(cWorld &) &) Line 1790
 	[Inline Frame] Cuberite.exe!std::invoke(cEntity::Destroy::__l2::void <lambda>(cWorld &) &) Line 1790
 	[Inline Frame] Cuberite.exe!std::_Invoker_ret<void,1>::_Call(cEntity::Destroy::__l2::void <lambda>(cWorld &) &) Line 1815
 	Cuberite.exe!std::_Func_impl_no_alloc<void <lambda>(cWorld &),void,cWorld & __ptr64>::_Do_call(cWorld & <_Args_0>) Line 299
 	[Inline Frame] Cuberite.exe!std::_Func_class<void,cWorld &>::operator()(cWorld &) Line 350
 	Cuberite.exe!cWorld::TickQueuedTasks() Line 1211
 	Cuberite.exe!cWorld::Tick(std::chrono::duration<__int64,std::ratio<1,1000> > a_Dt, std::chrono::duration<__int64,std::ratio<1,1000> > a_LastTickDurationMSec) Line 1025
 	Cuberite.exe!cWorld::cTickThread::Execute() Line 104
 	[Inline Frame] Cuberite.exe!std::_Invoker_pmf_pointer::_Call(void(cIsThread::*)() _Pmf, cIsThread * &&) Line 1790
 	[Inline Frame] Cuberite.exe!std::invoke(void(cIsThread::*)() &&) Line 1790
 	[Inline Frame] Cuberite.exe!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *>,std::default_delete<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *> > > >::_Execute(std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *> &) Line 240
 	[Inline Frame] Cuberite.exe!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *>,std::default_delete<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *> > > >::_Run(std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *>,std::default_delete<std::tuple<void (__cdecl cIsThread::*)(void),cIsThread *> > > > *) Line 247
 	Cuberite.exe!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl cIsThread::*)(void) __ptr64,cIsThread * __ptr64>,std::default_delete<std::tuple<void (__cdecl cIsThread::*)(void) __ptr64,cIsThread * __ptr64> > > >::_Go() Line 232
 	Cuberite.exe!std::_Pad::_Call_func(void * _Data) Line 211
 	[Inline Frame] Cuberite.exe!invoke_thread_procedure(unsigned int(*)(void *)) Line 91
 	Cuberite.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter) Line 115
```

**Comment by lkolbly** [comment]
I wonder if this is caused by the same root issue as #3971 (i.e. race condition calling ParentChunk->RemoveEntity and saving the world). Do you recall what you were doing when the server crashed? Also, does your server have an abnormally fast save cycle and/or destroy large numbers of entities right when the world is saving (or in general, really)?

**Comment by Seadragon91** [comment]
It's from APIFuzzing and yes there are done really quick calls. Will look if the pull request fixes it.

**Comment by Seadragon91** [comment]
Tested #4003 and the Verify still fails. If someone wants look at it. Get APIFuzzing, comment out this [line](https://github.com/cuberite/APIFuzzing/blob/45468536f6fa900680f2cd29e959e8558099bb10/tables.lua#L238) and run `checkapi` multiple times from console.

**Comment by Seadragon91** [comment]
This happens: The Entity is spawned and then directly Destroy is called. The entity is scheduled to be added to the world in the tick thread. That means for at least 1 tick the parent chunk is nullptr.

Example Code:
```
local world = cRoot:Get():GetDefaultWorld()
world:PrepareChunk(0, 0,
	function()
		local entityID = world:SpawnItemPickup(0, 100, 0, cItem(1, 64))
		world:DoWithEntityByID(entityID,
			function(a_Entity)
				a_Entity:Destroy()
				print("Destroyed")
			end)
	end)
```

**Comment by peterbell10** [comment]
So really there needs to be a `cWorld::RemoveEntity` that checks `m_EntitiesToAdd` if parent chunk is `nullptr`.  I'll try adding this to #4003.

**Comment by peterbell10** [comment]
Fixed by #4003
